pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStatus.java

```diff
@@ -43,17 +44,79 @@
  */
 @SuppressWarnings("unused")
 public class ServiceStatus {
+  public static final String STATUS_DESCRIPTION_NONE = "None";
+  public static final String STATUS_DESCRIPTION_INIT = "Init";
+  public static final String STATUS_DESCRIPTION_STARTED = "Started";
+  public static final String STATUS_DESCRIPTION_NO_HELIX_STATE = "Helix state does not exist";
   private static final Logger LOGGER = LoggerFactory.getLogger(ServiceStatus.class);
+  private static final int MAX_RESOURCE_NAMES_TO_LOG = 5;
+  private static final Map<String, ServiceStatusCallback> serviceStatusCallbackMap = new ConcurrentHashMap<>();
+  private static final ServiceStatusCallback serviceStatusCallback =
+      new MapBasedMultipleCallbackServiceStatusCallback(serviceStatusCallbackMap);
 
-  public enum Status {
-    STARTING, GOOD, BAD
+  public static void setServiceStatusCallback(String name, ServiceStatusCallback serviceStatusCallback) {
+    ServiceStatus.serviceStatusCallbackMap.put(name, serviceStatusCallback);
   }
 
-  public static final String STATUS_DESCRIPTION_NONE = "None";
-  public static final String STATUS_DESCRIPTION_INIT = "Init";
-  public static final String STATUS_DESCRIPTION_NO_HELIX_STATE = "Helix state does not exist";
+  public static void removeServiceStatusCallback(String name) {
+    ServiceStatus.serviceStatusCallbackMap.remove(name);
+  }
 
-  private static final int MAX_RESOURCE_NAMES_TO_LOG = 5;
+  public static Status getServiceStatus() {
+    return getServiceStatus(serviceStatusCallback);
+  }
+
+  public static Status getServiceStatus(String name) {
+    if (serviceStatusCallbackMap.containsKey(name)) {
+      return getServiceStatus(serviceStatusCallbackMap.get(name));
+    } else {
+      return Status.NOT_STARTED;
+    }
+  }
+
+  private static Status getServiceStatus(ServiceStatusCallback callback) {
+    try {
+      return callback.getServiceStatus();
+    } catch (Exception e) {
+      LOGGER.warn("Caught exception while reading the service status", e);
+      return Status.BAD;
+    }
+  }
+
+  public static String getStatusDescription() {
+    return getStatusDescription(serviceStatusCallback);
+  }
+
+  public static String getStatusDescription(String name) {
+    if (serviceStatusCallbackMap.containsKey(name)) {
+      return getStatusDescription(serviceStatusCallbackMap.get(name));
+    } else {
+      return STATUS_DESCRIPTION_NONE;
+    }
+  }
+
+  private static String getStatusDescription(ServiceStatusCallback callback) {
+    try {
+      return callback.getStatusDescription();
+    } catch (Exception e) {
+      return "Exception: " + e.getMessage();
```

Comment:
This should never happen. Can you add an error log here with the exception stack?
Also, please modify the return string to be `"Exception while getting service status: " + e.getMessage()`

Reply:
done